src/types-and-traits.rst

```diff
@@ -2182,16 +2184,70 @@ to :t:`[floating-point type]s`.
 A :t:`diverging type variable` is a :t:`type variable` that can refer to any
 :t:`type` and originates from a :t:`diverging expression`.
 
-:dp:`fls_zHpBXxfUWDz3`
-The :t:`expected type` of a :t:`let initializer` of a :t:`let statement` is
-determined as follows:
-
-* :dp:`fls_qob4wjgza3i8`
-  If the :t:`let statement` appears with a :t:`type ascription`, then the
-  :t:`expected type` is the :t:`type` specified by its :t:`type ascription`.
-
-* :dp:`fls_7vdr0mh7kmpz`
-  Otherwise the :t:`expected type` is a :t:`global type variable`.
+:dp:`fls_rvj3XspFZ1u3`
+The :t:`type inference` algorithm uses :t:`type unification` to propagate known
+:t:`[type]s` of :t:`[expression]s` and :t:`[pattern]s` across the
+:t:`type inference root` being inferred. In the rules detailed below, a static
+error occurs when :t:`type unification` fails.
+
+:dp:`fls_SZgixDCAx6PQ`
+:t:`Type inference` for a :t:`type inference root` proceeds as follows:
+
+#. :dp:`fls_XYY1U9h9HlAa`
+   Recursively process all :t:`[expression]s` and :t:`[statement]s` in the
+   :t:`type inference root` in program order.
```

Comment:
```suggestion
   :t:`type inference root` in declaration order.
```
We have used that term before.

Reply:
Good point!

Reply:
Actually "evaluation order" is the better term here, "program order" typically means something else I believe

Reply:
"Evaluation order" is a tricky one because in most cases it is "left-to-right", but in certain cases it is "right-to-left". Also, we have been using "declarative order" and "left-to-right order" in describing the dynamic semantics of evaluation and execution, so "evaluation order" seems to be a combination of "top-to-bottom" and "left-to-right" with the occasional "right-to-left".